test/gateway/index.js

```diff
@@ -0,0 +1,213 @@
+          content('nested-folder/ipfs.txt'),
+          content('nested-folder/nested.html'),
+          emptyDir('nested-folder/empty')
+        ]
+
+        http.api.node.files.add(dirs, (err, res) => {
+          expect(err).to.not.exist()
+          const root = res[res.length - 1]
+
+          expect(root.path).to.equal('test-folder')
+          expect(root.hash).to.equal(expectedRootMultihash)
+          cb()
+        })
+      },
+      (cb) => {
+        const expectedMultihash = 'Qme79tX2bViL26vNjPsF3DP1R9rMKMvnPYJiKTTKPrXJjq'
+
+        http.api.node.files.add(bigFile, (err, res) => {
+          expect(err).to.not.exist()
+          const file = res[0]
+          expect(file.path).to.equal(expectedMultihash)
+          expect(file.hash).to.equal(expectedMultihash)
+          cb()
+        })
+      }
+    ], done)
+  })
+
+  after((done) => http.api.stop(done))
+
+  describe('## HTTP Gateway', () => {
+    it('returns 400 for request without argument', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs'
+      }, (res) => {
+        expect(res.statusCode).to.equal(400)
+        expect(res.result.Message).to.be.a('string')
+        done()
+      })
+    })
+
+    it('400 for request with invalid argument', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/invalid'
+      }, (res) => {
+        expect(res.statusCode).to.equal(400)
+        expect(res.result.Message).to.be.a('string')
+        done()
+      })
+    })
+
+    it('valid hash', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/QmT78zSuBmuS4z925WZfrqQ1qHaJ56DQaTfyMUF7F8ff5o'
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.rawPayload).to.eql(Buffer.from('hello world' + '\n'))
+        expect(res.payload).to.equal('hello world' + '\n')
+        done()
+      })
+    })
+
+    it('stream a large file', (done) => {
+      let bigFileHash = 'Qme79tX2bViL26vNjPsF3DP1R9rMKMvnPYJiKTTKPrXJjq'
+
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/' + bigFileHash
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.rawPayload).to.eql(bigFile)
+        done()
+      })
+    })
+
+    it('load a non text file', (done) => {
+      let kitty = 'QmW2WQi7j6c7UgJTarActp7tDNikE4B2qXtFCfLPdsgaTQ/cat.jpg'
+
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/' + kitty
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.headers['content-type']).to.equal('image/jpeg')
+        done()
+      })
+    })
+
+    it('load a directory', (done) => {
+      let dir = 'QmW2WQi7j6c7UgJTarActp7tDNikE4B2qXtFCfLPdsgaTQ/'
+
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/' + dir
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')
```

Comment:
Should make some sort of assertion on the content here, maybe just a `indexOf` of the content for "Directory listing of"